### PR TITLE
Servlet 4.0 api tests

### PR DESCRIPTION
**Pull request by Rutzzzz**
https://rhoar-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/thorntail-2/job/thorntail-2-standalone-ts-rhel8-openjdk8/71/console

These tests only work for java 11, because of usage of new http/2 client api in java 11.

### CODE REVIEW

**Comment by Ladicek** [comment]
Using Apache HTTP Client like in other tests isn't possible?

**Comment by Ladicek** [comment]
There's a beta version of Apache HTTP Client that supports HTTP/2 (`org.apache.httpcomponents.client5:httpclient5:5.0-beta5` and `httpclient5-fluent`). I'd use that -- much better than only running on JDK 11.

**Comment by Ladicek** [comment]
Also, what about the files from `microprofile/microprofile-rest-client-1.3`? Are they really supposed to be in this PR?

File: javaee8/servlet-4.0/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        <relativePath>../..</relativePath>
+    </parent>
+
+    <artifactId>ts-javaee-servlet-4.0</artifactId>
+    <packaging>war</packaging>
+
+    <name>Thorntail TS: Java EE 8: Servlet 4.0</name>
+
+    <properties>
+        <failOnMissingWebXml>false</failOnMissingWebXml>
```
Comment (Ladicek) [inline javaee8/servlet-4.0/pom.xml]: This is already set in parent POM.

File: javaee8/servlet-4.0/pom.xml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <finalName>servlet-4.0</finalName>
```
Comment (Ladicek) [inline javaee8/servlet-4.0/pom.xml]: Not needed.

File: javaee8/servlet-4.0/pom.xml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    <execution>
+                        <goals>
+                            <goal>package</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
```
Comment (Ladicek) [inline javaee8/servlet-4.0/pom.xml]: Remove this entire plugin. If we _really_ need to run this on JDK 11 only, this test should be moved to the `java11` profile.

File: javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/MappingServletTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        sb.append("Mapping match: ");
+        sb.append(mappingMatch);
+        sb.append("\nServlet name: ");
+        sb.append(servletName);
+        sb.append("\nMatch value: ");
+        sb.append(matchingValue);
+        sb.append("\nPattern: ");
+        sb.append(pattern);
+        sb.append("\n");
+        return sb.toString();
```
Comment (Ladicek) [inline javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/MappingServletTest.java]: `"Foo: " + foo + "\nBar: " + bar` is _exactly_ the same as `StringBuilder sb = new StringBuilder(); sb.append("Foo: "); sb.append(foo); ... sb.toString()`. Resist the temptation ;-)

File: javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/PushServletTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "</head><body><p>secure</p><p>pushing succeeded</p></body></html>";
+    private static final String EXPECTED_FAILURE = "<html><head><title>Servlet 4.0</title>" +
+            "<link rel=\"stylesheet\" href=\"css/style.css\"/>" +
+            "</head><body><p>secure</p><p>pushing failed</p></body></html>";
+
+    @Test
+    @RunAsClient
+    public void serverPushSuccessTest() throws Exception {
+        String response = sendRequest("https://localhost:8443/PushServlet");
+        assertThat(response).isEqualTo(EXPECTED_SUCCESS);
```
Comment (Ladicek) [inline javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/PushServletTest.java]: Can you somehow assert that the CSS file was actually pushed?

**Comment by Ladicek** [comment]
Any comment about using the Apache HTTP Client 5 beta, instead of JDK 11 HTTP Client?

File: javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/MappingServletTest.java (hunk 4 above)
Comment (penehyba) [reply javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/MappingServletTest.java]: We discussed and disagree (among me and Ondra) what you meant by this comment.
He thinks you would prefer string concatenation, I think you would like chain of append().append().append.. and you just pointed out that separate calling to append has the same complexity. Please clarify (to me).

Comment (Ladicek) [reply javaee8/servlet-4.0/src/test/java/org/wildfly/swarm/ts/javaee8/servlet/MappingServletTest.java]: I most definitely prefer string concatenation of any unnecessary usage of `StringBuilder`.

**Comment by Ladicek** [comment]
Any comment about using the Apache HTTP Client 5 beta, instead of JDK 11 HTTP Client?

**Comment by Rutzzzz** [comment]
I have tried to run Apache HTTP Client 5 beta example at https://hc.apache.org/httpcomponents-core-5.0.x/examples.html [HTTP/2 requests over TLS connections with ALPN support on Java 1.7 and Java 1.8], but I could not get it to work.
Besides their api does not seem to support explicit receiving of push requests like the java11 does.

**Comment by penehyba** [comment]
LGTM

**Comment by Ladicek** [comment]
I played a little with the Apache HTTP Client 5 Beta and concluded that it's currently unnecessarily hard to use. Hopefully it will get better. For now, I agree with using the JDK 11 HTTP Client.

**Pull request merged**